src/dotnet/commands/dotnet-build/IncrementalManager.cs

```diff
@@ -0,0 +1,201 @@
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = InputItemsChanged(graphNode, compilerIO);
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = TimestampsChanged(compilerIO);
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            return IncrementalResult.DoesNotNeedRebuild;
+            
+        }
+
+        private IncrementalResult CLIChanged(ProjectGraphNode graphNode)
+        {
+            var currentVersionFile = DotnetFiles.VersionFile;
+            var versionFileFromLastCompile = graphNode.ProjectContext.GetSDKVersionFile(_configuration, _buildBasePath, _outputPath);
+
+            if (!File.Exists(currentVersionFile))
+            {
+                // this CLI does not have a version file; cannot tell if CLI changed
+                return IncrementalResult.DoesNotNeedRebuild;
+            }
+
+            if (!File.Exists(versionFileFromLastCompile))
+            {
+                // this is the first compilation; cannot tell if CLI changed
+                return IncrementalResult.DoesNotNeedRebuild;
+            }
+
+            var currentContent = DotnetFiles.ReadAndInterpretVersionFile();
+
+            var versionsAreEqual = string.Equals(currentContent, File.ReadAllText(versionFileFromLastCompile), StringComparison.OrdinalIgnoreCase);
+
+            return versionsAreEqual
+                ? IncrementalResult.DoesNotNeedRebuild
+                : new IncrementalResult("the version or bitness of the CLI changed since the last build");
+        }
+
+        private IncrementalResult InputItemsChanged(ProjectGraphNode graphNode, CompilerIO compilerIO)
+        {
+            // check empty inputs / outputs
+            if (!compilerIO.Inputs.Any())
+            {
+                return new IncrementalResult("the project has no inputs");
+            }
+
+            if (!compilerIO.Outputs.Any())
+            {
+                return new IncrementalResult("the project has no outputs");
+            }
+
+            // check non existent items
+            var result = CheckMissingIO(compilerIO.Inputs, "inputs");
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = CheckMissingIO(compilerIO.Outputs, "outputs");
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            // check cache against input glob pattern changes
+            var incrementalCacheFile = graphNode.ProjectContext.IncrementalCacheFile(_configuration, _buildBasePath, _outputPath);
+
+            if (!File.Exists(incrementalCacheFile))
+            {
+                // no cache present; cannot tell if anything changed
+                return IncrementalResult.DoesNotNeedRebuild;
```

Comment:
Quite a good question. I was also not very happy with the decision, and see pros and cons for either way.

If it _did_ trigger a rebuild when the cache file is absent, it would mean that the incremental manager is implicitly coupled to the precondition that some other piece of code generates it.
When that precondition gets broken and no other piece of code generates that file, then the IncrementalManager will always answer with rebuilding, killing incrementality off. I did not like that implicit assumption.

If it _does not_ trigger a rebuild when the cache file is absent, it puts the responsibility in the hands of the caller, and if the calling code decides not to generate the file, then incremental proceeds with the other heuristics. 

First option is error proof, choosing to be conservative, but kills all incrementality. Second option may lead to false negative builds (turns the feature that this PR introduces off).

There's pros and cons for each side. If you still think it should trigger a rebuild, then I'll change it. The same decision is taken for the .SDKVersion file as well, so I'll change it there as well.